celery/canvas.py

```diff
@@ -501,7 +501,10 @@ def stamp(self, visitor=None, **headers):
         """
         headers = headers.copy()
         if visitor is not None:
-            headers.update(visitor.on_signature(self, **headers))
+            visitor_headers = visitor.on_signature(self, **headers)
+            if "stamped_headers" not in visitor_headers:
```

Comment:
additional unit test will be useful for this part

Reply:
Done.

Reply:
let me recheck

Reply:
Feel free to merge if it passes your code review @auvipy